lib/concurrent/parallel.rb

```diff
@@ -0,0 +1,91 @@
+require 'concurrent/atomics'
+require 'concurrent/configuration'
+require 'concurrent/executors'
+
+module Concurrent
+
+  class Parallel < SimpleDelegator
+    module EnumerableInterface
+      module_function
+
+      def use(runner, methods)
+        enumerable_methods = Enumerable.instance_methods.map(&:to_s)
+        enumerable_methods << 'each'
+        methods.each do |method|
+          next unless enumerable_methods.include?(method)
+          class_eval <<-RUBY
+          def #{method}(*args, &block)
+            #{runner}(:#{method}, *args, &block)
+          end
+          RUBY
+        end
+      end
+
+      use :run_in_threads, %w(
+        all? any? count detect find find_index max_by min_by minmax_by none?
+        one? partition
+      )
+
+      use :run_in_threads_return_original, %w(
+        cycle each each_cons each_entry each_slice each_with_index enum_cons
+        enum_slice enum_with_index reverse_each zip
+      )
+
+      use :run_in_threads_return_parallel, %w(
```

Comment:
@SebastianEdwards I don't understand the use of the `run_in_threads_return_parallel` and `run_in_threads_return_original` methods. The `collect` method, for example, should return a new array. If I understand this implementation correctly, won't a call to `Parallel.collect` return another parallel rather than an array?

As I think about this a little more I think I see where you are going. Rather than return an array we return an array-like object (because of the Delegator). The returned object acts just like an array, but all following operations on the returned array will be in parallel. Is this correct?

If my understanding is correct then this is definitely an ingenious implementation. I can definitely see the value, but I can also see this leading to some confusion. Least astonishment again. Now we're talking about explicit versus implicit behavior.

If a parallelized `collect` method returns a delegator then we get *implicit* parallelization of subsequent operations:

``ruby
[1, 2, 3].parallel.collect{|x| x * 2}.collect{|y| y * 3}
```

In this case *both* collect calls are parallel operations. The second call is implicitly parallel. The intent of the code isn't explicit, it must be inferred that the first parallel begets a second parallel call.

If, instead, we return a regular array from the first `collect` call it looks like this:

``ruby
[1, 2, 3].parallel.collect{|x| x * 2}.parallel.collect{|y| y * 3}
```

This example takes more typing but the parallelization of the second call is explicit. There is no ambiguity as to the intent of the code.

So, again, I'm a little on the fence. Personally, I prefer extra typing and explicit code. So my inclination would be to the second example. But I also see the value in the first example. I'm curious to hear what others on the team think.